### ISSUE TITLE
[NoXtend] We need a java-based Formatter2 API

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext#1697]

**Comment by miklossy** [comment]
See also https://bugs.eclipse.org/bugs/show_bug.cgi?id=481563

**Comment by miklossy** [comment]
See also https://bugs.eclipse.org/bugs/show_bug.cgi?id=564265 and https://bugs.eclipse.org/bugs/show_bug.cgi?id=563046

CC @ewillink

**Comment by ArneDeutsch** [comment]
To prototype this I've implemented a simple grammar
```
grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
Model:
	parents+=Parent*;
Parent:
	'parent' name=ID '!' ('{' children+=Child* '}')?;
Child:
	'child' name=ID;
```
together with an typical Xtend formatter.
```
import static org.xtext.example.mydsl.myDsl.MyDslPackage.Literals.*

class MyDslFormatter extends AbstractFormatter2 {
	def dispatch void format(Model model, extension IFormattableDocument document) {
		for (parent : model.parents)
			parent.format
	}

	def dispatch void format(Parent parent, extension IFormattableDocument document) {
		parent.regionFor.keyword("parent").prepend[noSpace].append[oneSpace]
		parent.regionFor.feature(PARENT__NAME).append[oneSpace]
		parent.regionFor.keyword("{").prepend[oneSpace].append[newLine]
		parent.regionFor.keyword("}").append[setNewLines(1, 1, 2)]
		interior(parent.regionFor.keyword("{"), parent.regionFor.keyword("}"))[indent]
		for (child : parent.children)
			child.format
	}

	def dispatch void format(Child child, extension IFormattableDocument document) {
		child.regionFor.keyword("child").append[oneSpace]
		child.regionFor.feature(CHILD__NAME).append[newLine]
	}
}
```
Then I translated the generated class to as simple java I can think of. At last I extracted a new abstract class with an reflective `format` method and delegate methods for `ITextRegionExtensions`, Using this base class the resulting java formatter is not much more verbose as the Xtend one (it's still java, I'm afraid).
```
import static org.xtext.example.mydsl.myDsl.MyDslPackage.Literals.*;

public class MyDslFormatter2 extends AbstractJavaFormatter {
	protected void _format(Model model, IFormattableDocument doc) {
		for (Parent parent : model.getParents())
			doc.format(parent);
	}

	protected void _format(Parent parent, IFormattableDocument doc) {
		doc.prepend(regionFor(parent).keyword("parent"), it -> it.noSpace());
		doc.append(regionFor(parent).keyword("parent"), it -> it.oneSpace());
		doc.append(regionFor(parent).feature(PARENT__NAME), it -> it.oneSpace());
		doc.prepend(regionFor(parent).keyword("{"), it -> it.oneSpace());
		doc.append(regionFor(parent).keyword("{"), it -> it.newLine());
		doc.interior(regionFor(parent).keyword("{"), regionFor(parent).keyword("}"), it -> it.indent());
		doc.append(regionFor(parent).keyword("}"), it -> it.setNewLines(1, 1, 2));
		for (Child child : parent.getChildren())
			doc.format(child);
	}

	protected void _format(Child child, IFormattableDocument doc) {
		doc.append(regionFor(child).keyword("child"), it -> it.oneSpace());
		doc.append(regionFor(child).feature(CHILD__NAME), it -> it.newLine());
	}
}
```
For completness the abstract base class:
```
public abstract class AbstractJavaFormatter extends AbstractFormatter2 implements ITextRegionExtensions {

	@Override
	public void format(Object child, IFormattableDocument document) {
		if (child instanceof XtextResource) {
			_format((XtextResource) child, document);
			return;
		} else if (child == null) {
			_format((Void) null, document);
			return;
		}
		try {
			Method[] methods = getClass().getDeclaredMethods();
			for (Method _format : methods) {
				if (!_format.getName().equals("_format"))
					continue;
				if (_format.getParameterCount() != 2)
					continue;
				if (_format.getParameters()[1].getType() != IFormattableDocument.class)
					continue;
				if (!_format.getParameters()[0].getType().isInstance(child))
					continue;
				_format.invoke(this, child, document);
				return;
			}
			if (child instanceof EObject) {
				_format((EObject) child, document);
				return;
			} else if (child != null) {
				_format(child, document);
				return;
			} else {
				throw new IllegalArgumentException(
						"Unhandled parameter types: " + Arrays.asList(child, document).toString());
			}
		} catch (SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
			throw new RuntimeException(e);
		}
	}

	@Override
	public ISemanticRegionsFinder allRegionsFor(EObject semanticElement) {
		return textRegionExtensions.allRegionsFor(semanticElement);
	}

	@Override
	public Iterable<ISemanticRegion> allSemanticRegions(EObject semanticElement) {
		return textRegionExtensions.allSemanticRegions(semanticElement);
	}

	@Override
	public EObject grammarElement(EObject semanticElement) {
		return textRegionExtensions.grammarElement(semanticElement);
	}

	@Override
	public ISemanticRegionFinder immediatelyFollowing(EObject semanticElement) {
		return textRegionExtensions.immediatelyFollowing(semanticElement);
	}

	@Override
	public ISemanticRegionFinder immediatelyPreceding(EObject semanticElement) {
		return textRegionExtensions.immediatelyPreceding(semanticElement);
	}

	@Override
	public boolean isMultiline(EObject semanticElement) {
		return textRegionExtensions.isMultiline(semanticElement);
	}

	@Override
	public IHiddenRegion nextHiddenRegion(EObject semanticElement) {
		return textRegionExtensions.nextHiddenRegion(semanticElement);
	}

	@Override
	public IHiddenRegion previousHiddenRegion(EObject semanticElement) {
		return textRegionExtensions.previousHiddenRegion(semanticElement);
	}

	@Override
	public ISemanticRegionsFinder regionFor(EObject semanticElement) {
		return textRegionExtensions.regionFor(semanticElement);
	}

	@Override
	public IEObjectRegion regionForEObject(EObject semanticElement) {
		return textRegionExtensions.regionForEObject(semanticElement);
	}

	@Override
	public Iterable<ISemanticRegion> semanticRegions(EObject semanticElement) {
		return textRegionExtensions.semanticRegions(semanticElement);
	}
}
```
What do you think? Simple enough? Probably more difficult cases I havn't seen?

**Comment by ewillink** [comment]
I have elaborated https://bugs.eclipse.org/bugs/show_bug.cgi?id=564254 with what I see as the real problem that the above does not address.

**Comment by ArneDeutsch** [comment]
If I understand correcly you refere to the warnings that the API is not final. Correct? In fact this is not nice but I do not understand why you name it as "the real problem". From my point of view we could mark the API final. There was no change over the last years to the API (it is stable), and it is quite usable. No intention to change it. But of course I might miss reasons why the API is not arked as public. Anyone?

**Comment by ewillink** [comment]
The APIs from early Xtext development were made public; good for users - painful for developers.

Subsequent development of "2" variants used failed to export packages so that users are faced with not using the facilitoes at all or just ignoring numerous warnings.

Please remove the privileged status of Xtext, the self-consumer, as an X-friend of Xtext, the producer.

**Comment by cdietrich** [comment]
TODO: add some hints to release notes. point to samples in xtext-core, xtext-web and the XtextFormatter java

**Comment by cdietrich** [comment]
ping @ArneDeutsch (just to make sure this does not get lost)

**Comment by cdietrich** [comment]
Fixed in 2.24
